File: aspect-java/src/main/java/com/pedrocomitto/poc/aspect/service/BusinessLogicService.java
package com.pedrocomitto.poc.aspect.service;

import com.pedrocomitto.poc.aspect.aop.annotation.Idempotent;
import org.springframework.stereotype.Service;

@Service
public class BusinessLogicService {

    @Idempotent
    public void process() {
        System.out.println("business logic");
    }

}

File: aspect-kotlin/src/main/kotlin/com/pedrocomitto/poc/aspect/service/BusinessLogicService.kt
package com.pedrocomitto.poc.aspect.service

import com.pedrocomitto.poc.aspect.aop.withIdempotency
import org.springframework.stereotype.Service

@Service
class BusinessLogicService {

    fun process() {
        withIdempotency { println("business logic") }
    }

}
